src/gather_classes.rs

```diff
@@ -183,42 +195,42 @@ fn process_compare_paths(
             100.00 * ((value as f64) / (total as f64)),
         )
     };
-    let mut failures = 0;
-    let mut processed = 0;
 
     let mut overall_report = overall_report;
-    // TODO (#72) Potential speedup: Process each file with `Process(path) -> ClassHashesReport` in parallel, then merge ClassHashesReport together.
-    for (block_num, path) in file_paths.into_iter() {
-        info!(
-            "{}\t{}\t{}",
-            get_stat("Processed", processed),
-            get_stat("Successes", processed - failures),
-            get_stat("Failures", failures)
-        );
-        info!(
-            "Processing path: `{}`",
-            path.file_name().unwrap().to_str().unwrap()
-        );
-        processed += 1;
+    let semaphore = Arc::new(Semaphore::new(max_block_processing_count));
+
+    let mut join_set = JoinSet::new();
+    for (block_num, path) in file_paths {
+        let sempahore_clone = Arc::clone(&semaphore);
+        join_set.spawn(async move {
+            let _permit = sempahore_clone.acquire().await.unwrap();
+            process_single_file(block_num, path).await
+        });
+    }
 
-        let json_result: Result<Value, _> = try_deserialize(path);
-        if let Err(err) = json_result {
-            warn!("Failed to process file with err: {err}");
-            failures += 1;
-            continue;
-        }
-        let json_obj = json_result.unwrap();
+    let mut processed = 0;
+    let mut failures = 0;
 
-        let update_result = update_report(&mut overall_report, &json_obj, block_num);
-        match update_result {
```

Comment:
There is no need to handle the async process. Let's use an async library which has all of this already implemented:

* https://docs.rs/futures/latest/futures/

The idea is that you tell him what to execute and it will return the values as they get calculated. You don't have to manually select them.

Reply:
I'm not sure I understand, you mean something like this?

```
let futures: FuturesUnordered<_> = file_paths
        .into_iter()
        .map(|(block_num, path)| {
            let semaphore_clone = Arc::clone(&semaphore);
            async move {
                let _permit = semaphore_clone.acquire().await.unwrap();
                process_single_file(block_num, path).await
            }
        })
        .collect();
        
        futures.for_each(|result| async { ... });
```

Reply:
There is no need for a semaphore as far as I understand. You scan each files at the "same" time, once they've done scanning, they will be returned one by one, depending on who finished first. The only thing you have to do is _await_ for them to finish and process the result. 

The result processing needs to be done synchronously. Again, you can abstract yourself away and remove the existence of a semaphore.

1. First iteration scans all the files
2. Second interation awaits for results and process at the same time

Am I missing something?

Reply:
I thought the goal was to have X files to be processed in parallel, so once 1 of the file is done processing, then the next one can continue. Then the JoinSet makes it easy to wait until some file is done processing to update the final json, without race condition